Add self and remote attach capabilities (#420)
elastic/apm-agent-java: docs/attach-setup.asciidoc

[[attach-setup]]
== Setup the agent with the attacher

You can use the attacher as a standalone application on the command line or include it in your application to programmatically attach to the current JVM.

This installation method does not require you to alter the configuration of your application server and can be used to conveniently instrument all JVMs on a particular host.

> NOTE: This installation method is experimental.

Reviewer: I think this note should be above these two paragraphs at the top of the page

[float]
[[attach-setup-cli]]
=== Use the attacher as a command line application

The `apm-agent-attach.jar` is a small Java program which attaches the Elastic APM Java agent to a specific JVM or to all JVMs of the same host it runs on.

> This requires that the OS user of the attach is the same as the OS user of the applications to be monitored (the JVMs to attach to).

Reviewer: This paragraph was a little confusing to me, but it might just be because I'm not familiar with some of the terminology. I'm wondering if this would be better:

```suggestion
This requires that the OS user of the attach and the OS user of the application(s) to be monitored (the JVMs to attach to) are the same.
```

Author: That sentence is indeed confusing. I was trying to say the OS user used to execute the apm-agent-attach.jar and the JVMs the agent should be attached to have to match.

[float]
[[attach-setup-cli-download]]
=== Download

You can download the attach program from maven central:

> link:https://search.maven.org/search?q=g:co.elastic.apm%20AND%20a:apm-agent-attach[maven central]

Reviewer: Should this link work yet? It just brings me to a blank search page

Author: That link will only work once this change gets released.

[float]
[[attach-setup-usage-cli]]
=== Usage

[float]
[[attach-setup-usage-cli-pid]]
==== Attach to a JVM with a specific PID

Attaches the agent to a JVM with a specific process ID.
Optionally, you can pass in arguments to configure the agent.
See <<attach-setup-usage-cli-options>> for the format of the agent arguments.

[source,bash]
----
java -jar apm-agent-attach.jar --pid <pid> [--args <agent_arguments>]
----

Example: The following command attaches the agent to the JVM with the PID 42 and exits.
Additionally, it applies some <<configuration,configuration options>>.

[source,bash]
----
java -jar apm-agent-attach.jar --pid 42 \
     --args 'service_name=my-cool-service;server_urls=http://localhost:8200'
----

[float]
[[attach-setup-usage-cli-filtered]]
==== Attach to a filtered set of JVMs on a certain host

> Attaches to agent to all matching JVMs on the host which confirm to the optional include and exclude filters.

Reviewer: ```suggestion
Attaches the agent to all matching JVMs on the host which confirm to the optional include and exclude filters.
```

       [--args <agent_arguments> | --args-provider <args_provider_script>]
----

Example: The following command attaches the agent to all JVMs whose main class contains `MyApplication` or which are started from a jar file named `my-application.jar`.
It also makes the attacher run continuously so that it attaches the agent on starting JVMs which match the include pattern.
Additionally, it applies some <<configuration,configuration options>>.

[source,bash]
----
java -jar apm-agent-attach.jar --include '.*MyApplication.*' '.*/my-application.jar' \
     --continuous \
     --args 'service_name=my-cool-service;server_urls=http://localhost:8200'
----

[float]
[[attach-setup-usage-cli-list]]
==== List running JVMs

Lists all currently running JVMs, including their PID and their main class name or the path to their jar file.

[source,bash]
----
java -jar apm-agent-attach.jar --list
----

[float]
[[attach-setup-usage-cli-options]]
==== Options

*-l, --list*::
+
--
Lists all running JVMs. Same output as `jps -l`.
--

*-p, --pid <pid>*::
+
--
PID of the JVM to attach. If not provided, attaches to all currently running JVMs which match the `--exclude` and `--include` filters.

> NOTE: this option can not be used in conjunction with `--continuous`, `--exclude`, `--include` and `--args-provider`

Reviewer: ```suggestion
NOTE: this option cannot be used in conjunction with `--continuous`, `--exclude`, `--include` and `--args-provider`
```

--

*-c, --continuous*::
+
--
If provided, this program continuously runs and attaches to all running and starting JVMs which match the `--exclude` and `--include` filters.

> NOTE: this option can not be used in conjunction with `--pid`

Reviewer: ```suggestion
NOTE: this option cannot be used in conjunction with `--pid`
```

--

*-e, --exclude <exclude_pattern>...*::
+
--
A list of regular expressions of fully qualified main class names or paths to JARs of applications the java agent should not be attached to.
(Matches the output of `jps -l`)

> NOTE: this option can not be used in conjunction with `--pid`

Reviewer: ```suggestion
NOTE: this option cannot be used in conjunction with `--pid`
```

--

*-i, --include <include_pattern>...*::
+
--
A list of regular expressions of fully qualified main class names or paths to JARs of applications the java agent should be attached to.
(Matches the output of `jps -l`)

> NOTE: this option can not be used in conjunction with `--pid`

Reviewer: ```suggestion
NOTE: this option cannot be used in conjunction with `--pid`
```

--

*-a, --args <agent_arguments>*::
+
--
If set, the arguments are used to configure the agent on the attached JVM (agentArguments of agentmain).

The syntax of the arguments is `key1=value1;key2=value1,value2`.
See <<configuration>> for all available configuration options.

> NOTE: this option can not be used in conjunction with `--args-provider`

Reviewer: ```suggestion
NOTE: this option cannot be used in conjunction with `--args-provider`
```

--

*-A, --args-provider <args_provider_script>*::
+
--
The name of a program which is called when a new JVM starts up.
The program gets the pid and the main class name or path to the JAR file as an argument
and returns an arg string which is used to configure the agent on the attached JVM (agentArguments of agentmain).
When returning a non-zero status code from this program, the agent will not be attached to the starting JVM.

The syntax of the arguments is `key1=value1;key2=value1,value2`.
See <<configuration>> for all available configuration options.

> NOTE: this option can not be used in conjunction with `--pid` and `--args`

Reviewer: ```suggestion
NOTE: this option cannot be used in conjunction with `--pid` and `--args`
```
